scylla/tests/ccm-integration/cluster.rs

```diff
@@ -0,0 +1,832 @@
+
+struct NodeListIter<'a> {
+    list: &'a NodeList,
+    index: usize,
+}
+
+impl<'a> Iterator for NodeListIter<'a> {
+    type Item = Arc<RwLock<Node>>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index < self.list.0.len() {
+            let node = self.list.0[self.index].clone();
+            self.index += 1;
+            Some(node)
+        } else {
+            None
+        }
+    }
+}
+
+impl NodeList {
+    fn push(&mut self, node: Arc<RwLock<Node>>) {
+        self.0.push(node);
+    }
+
+    fn iter(&self) -> NodeListIter {
+        NodeListIter {
+            list: self,
+            index: 0,
+        }
+    }
+
+    fn new() -> Self {
+        NodeList(Vec::new())
+    }
+
+    pub(crate) async fn get_contact_endpoints(&self) -> Vec<String> {
+        let mut out: Vec<String> = Vec::new();
+        for node in self.iter() {
+            let node = node.read().await;
+            let cp = node.contact_endpoint();
+            out.push(cp);
+        }
+        out
+    }
+}
+
+pub(crate) struct Cluster {
+    pub(crate) nodes: NodeList,
+    destroyed: bool,
+    logged_cmd: Arc<LoggedCmd>,
+    opts: ClusterOptions,
+}
+
+impl Drop for Cluster {
+    fn drop(&mut self) {
+        if !self.opts.do_not_remove_on_drop {
+            self.destroy_sync().ok();
+        }
+    }
+}
+
+pub(crate) const DEFAULT_MEMORY: u32 = 512;
+pub(crate) const DEFAULT_SMP: u16 = 1;
+
+impl Cluster {
+    // This method looks at all busy /24 networks in 127.0.0.0/8 and return first available
+    // network goes as busy if anything is listening on any port and any address in this network
+    // 127.0.0.0/24 is skipped and never returned
+    async fn sniff_ipv4_prefix() -> Result<NetPrefix, Error> {
+        let mut used_ips: HashSet<IpAddr> = HashSet::new();
+        let file = File::open("/proc/net/tcp").await?;
+        let mut lines = BufReader::new(file).lines();
+        while let Some(line) = lines.next_line().await? {
+            let parts: Vec<&str> = line.split_whitespace().collect();
+            if let Some(ip_hex) = parts.get(1) {
+                let ip_port: Vec<&str> = ip_hex.split(':').collect();
+                if let Some(ip_hex) = ip_port.get(0) {
+                    if let Some(ip) = u32::from_str_radix(ip_hex, 16).ok() {
+                        used_ips.insert(
+                            Ipv4Addr::new(ip as u8, (ip >> 8) as u8, (ip >> 16) as u8, 0).into(),
+                        );
+                    }
+                }
+            }
+        }
+
+        for a in 0..=255 {
+            for b in 0..=255 {
+                if a == 0 && b == 0 {
+                    continue;
+                }
+                let ip_prefix: IpAddr = Ipv4Addr::new(127, a, b, 0).into();
+                if !used_ips.contains(&ip_prefix) {
+                    return Ok(ip_prefix.into());
+                }
+            }
+        }
+        Err(Error::msg("All ip ranges are busy"))
+    }
```

Comment:
Its absurd that we have to do this, it should be ccm's job to find a free ips for the nodes.

Reply:
well, it does not do that and we have to dial with it.